saleor/payment/utils.py

```diff
@@ -125,9 +147,91 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
+        func_name, transaction_kind, payment, payment_token, **extra_params):
+    """Helper that calls the passed gateway function and handles exceptions.
+
+    Additionally does validation of the returned gateway response."""
+    gateway, gateway_params = get_payment_gateway(payment.gateway)
+    gateway_response = None
+    error_msg = None
+
+    payment_information = create_payment_information(
+        payment, payment_token, **extra_params
+    )
+
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment_information=payment_information, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        error_msg = 'Gateway doesn\'t implement {}'.format(func_name)
+        logger.exception(error_msg)
+    except GatewayError:
+        error_msg = 'Gateway response validation failed'
+        logger.exception(error_msg)
+        gateway_response = None  # set response empty as the validation failed
+    except Exception:
+        error_msg = 'Gateway encountered an error'
+        logger.exception(error_msg)
+    finally:
+        if not isinstance(gateway_response, list):
+            gateway_response = [gateway_response]
+        for response in gateway_response:
+            transaction = create_transaction(
+                payment=payment,
+                kind=transaction_kind,
```

Comment:
The ```kind``` is no longer necessary since the gateway response has to return the ```kind``` information as a required field.

Reply:
It is required, well.. kinda, this is in case a rare situation happens that gateway explodes with an exception and doesn't return a response, so we still create a transaction of a specific type.

Reply:
Makes sense!